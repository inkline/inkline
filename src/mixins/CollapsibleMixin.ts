import { defineComponent } from 'vue';
import { on, off } from '@inkline/inkline/helpers';
import { breakpoints } from '@inkline/inkline/constants';
import { Classes } from '@inkline/inkline/types';

export default defineComponent({
    props: {
        collapse: {
            type: [String, Boolean],
            default: 'md'
        },
        modelValue: {
            type: Boolean,
            default: false
        }
    },
    emits: ['update:modelValue'],
    data () {
        return {
            open: this.modelValue,
            windowWidth: typeof window !== 'undefined' ? window.innerWidth : 0
        };
    },
    computed: {
        collapsibleClasses (): Classes {
            return {
                '-open': this.open,
                '-collapsible': this.collapsible,
                [`-collapse-${this.collapse}`]: Boolean(this.collapse)
            };
        },
        collapsible (): boolean {
            if (this.collapse === true || this.collapse === false) {
                return this.collapse;
            }

            return this.windowWidth <= breakpoints[this.collapse][1];
        }
    },
    watch: {
        modelValue (value) {
            this.open = value;
        }
    },
    created () {
        if (typeof window !== 'undefined') {
            on(window as any, 'resize', this.onWindowResize);

            this.onWindowResize();
        }
    },
    beforeUnmount () {
        if (typeof window !== 'undefined') {
            off(window as any, 'resize', this.onWindowResize);
        }
    },
    methods: {
        setOpen (value: boolean) {
            this.open = value;
            this.$emit('update:modelValue', this.open);
        },
        toggleOpen () {
            this.open = !this.open;
            this.$emit('update:modelValue', this.open);
        },
        onWindowResize () {
            if (this.collapse === true || this.collapse === false) {
                return;
            }

            const windowWidth = window.innerWidth;

            if (this.windowWidth <= breakpoints[this.collapse][1] && windowWidth > breakpoints[this.collapse][1]) {
                this.setOpen(false);
            }

            this.windowWidth = window.innerWidth;
        }
    }
});
