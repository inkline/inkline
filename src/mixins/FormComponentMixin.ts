import { defineComponent } from 'vue';
import { getValueByPath } from '@inkline/inkline/helpers';

export default defineComponent({
    inject: {
        formGroup: {
            default: (): any => ({})
        },
        form: {
            default: (): any => ({})
        }
    },
    computed: {
        isDisabled(): boolean {
            return (this as any).disabled || (this as any).form.isDisabled || (this as any).formGroup.isDisabled;
        },
        isReadonly(): boolean {
            return (this as any).readonly || (this as any).form.isReadonly || (this as any).formGroup.isReadonly;
        },
        parent(): any {
            if ((this as any).formGroup.$) {
                return (this as any).formGroup;
            }

            return (this as any).form;
        },
        schema(): any {
            if ((this as any).name !== '') {
                return getValueByPath((this as any).parent.schema || {}, `${(this as any).name}`);
            }

            return (this as any).parent.schema || {};
        }
    }
});
