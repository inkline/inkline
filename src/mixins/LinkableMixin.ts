import { defineComponent } from 'vue';

export default defineComponent({
    props: {
        tag: {
            type: String,
            default: 'a'
        }
    },
    computed: {
        isTag (): string {
            return this.$attrs.to ? this.routerComponent : this.$attrs.href ? 'a' : this.tag;
        },
        isComponent (): boolean {
            return this.isTag === this.routerComponent;
        },
        routerComponent (): string {
            return this.$inkline.options.routerComponent;
        }
    }
});
