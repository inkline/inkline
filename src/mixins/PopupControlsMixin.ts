import { defineComponent } from 'vue';
import { on, off, focusFirstDescendant } from '@inkline/inkline/helpers';

export default defineComponent({
    props: {
        disabled: {
            type: Boolean,
            default: false
        },
        modelValue: {
            type: Boolean,
            default: undefined
        },
        trigger: {
            type: Array,
            default: () => ['hover', 'click', 'focus']
        }
    },
    emits: ['update:modelValue'],
    data() {
        return {
            visible: this.modelValue,
            triggerStack: 0
        };
    },
    watch: {
        modelValue(value) {
            if (value) {
                this.show();
            } else {
                this.hide();
            }
        }
    },
    mounted() {
        if (((this as any).$slots.default() || []).length === 0) {
            throw new Error('Popup components require one child element to be used as trigger.');
        }

        this.addEventListeners();
    },
    beforeUnmount() {
        this.removeEventListeners();
    },
    methods: {
        show() {
            if (this.disabled || this.visible) {
                return;
            }

            this.triggerStack += 1;
            this.visible = true;

            (this as any).createPopper();
            this.$emit('update:modelValue', true);
        },
        hide() {
            if (this.disabled || !this.visible) {
                return;
            }

            this.triggerStack -= 1;

            if (this.triggerStack <= 0) {
                this.triggerStack = 0;
                this.visible = false;

                // Destroyed by <transition> component after transition is finished
                // this.destroyPopper();
                this.$emit('update:modelValue', false);
            }
        },
        onClick() {
            if (this.visible) {
                this.hide();
            } else {
                this.show();
            }
        },
        onClickOutside() {
            if (this.modelValue) return;

            this.hide();
        },
        addEventListeners() {
            [].concat((this as any).trigger).forEach((trigger) => {
                switch (trigger) {
                case 'hover':
                    on(this.$refs.trigger as HTMLElement, 'mouseenter', this.show);
                    on(this.$refs.trigger as HTMLElement, 'mouseleave', this.hide);
                    break;
                case 'click':
                    on(this.$refs.trigger as HTMLElement, 'click', this.onClick);
                    break;
                case 'focus':
                    for (const child of (this as any).$refs.trigger.children) {
                        on(child, 'focus', this.show);
                        on(child, 'blur', this.hide);
                    }
                    break;
                default:
                    break;
                }
            });
        },
        removeEventListeners() {
            [].concat((this as any).trigger).forEach((trigger) => {
                switch (trigger) {
                case 'hover':
                    off(this.$refs.trigger as HTMLElement, 'mouseenter', this.show);
                    off(this.$refs.trigger as HTMLElement, 'mouseleave', this.hide);
                    break;
                case 'click':
                    off(this.$refs.trigger as HTMLElement, 'click', this.onClick);
                    break;
                case 'focus':
                    for (const child of (this as any).$refs.trigger.children) {
                        off(child, 'focus', this.show);
                        off(child, 'blur', this.hide);
                    }
                    break;
                default:
                    break;
                }
            });
        },
        focusTrigger() {
            for (const child of (this as any).$refs.trigger.children) {
                if (focusFirstDescendant(child)) {
                    child.focus();
                    break;
                }
            }
        }
    }
});
