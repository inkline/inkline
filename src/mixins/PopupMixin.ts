import { defineComponent } from 'vue';
import { createPopper, Instance, Placement } from '@popperjs/core';

export const offsetModifier = (offset: number) => ({
    name: 'offset',
    options: {
        offset: [0, offset]
    }
});

export const arrowModifier = () => ({
    name: 'arrow',
    options: {
        padding: 6 // padding from the edges of the popper
    }
});

export const preventOverflowModifier = () => ({
    name: 'preventOverflow',
    options: {
        padding: 8 // padding from the edges of the viewport
    }
});

export const computeStylesModifier = () => ({
    name: 'computeStyles',
    options: {
        gpuAcceleration: false,
        adaptive: false // true by default
    }
});

export const sameWidthModifier = () => ({
    name: 'sameWidth',
    enabled: true,
    phase: 'beforeWrite',
    requires: ['computeStyles'],
    fn: ({ state }: { state: any }) => {
        state.styles.popper.width = `${state.rects.reference.width}px`;
    },
    effect ({ state }: { state: any }) {
        state.elements.popper.style.width = `${state.elements.reference.offsetWidth}px`;
    }
});

export const useBaseModifiers = ({ offset }: { offset: number }) => [
    offsetModifier(offset),
    arrowModifier(),
    preventOverflowModifier(),
    computeStylesModifier()
];

/**
 * @param {HTMLElement} [popupElement=$refs.overlay]
 * The HTML element used as overlay, or a configuration used to generate the overlay.
 * @param {String} [placement=auto]
 * Placement of the overlay, accepted values:
 *  - auto(-start, -end)
 *  - top(-start, -end)
 *  - right(-start, -end)
 *  - bottom(-start, -end)
 *  - left(-start, -end)
 * @param {Boolean} [visible=false]
 * Visibility of the overlay
 * @param {Number} [offset=0]
 * Amount of pixels the overlay will be shifted (can be negative).
 */

export default defineComponent({
    props: {
        placement: {
            type: String,
            default: 'auto' as Placement
        },
        offset: {
            type: Number,
            default: 6
        },
        popperOptions: {
            type: Object,
            default: () => ({})
        }
    },
    data (): { popperInstance?: Instance } {
        return {
            popperInstance: undefined
        };
    },
    watch: {
        placement (placement: Placement) {
            if (this.popperInstance) {
                this.popperInstance.setOptions({ placement });
            }
        }
    },
    beforeUnmount () {
        this.destroyPopper();
    },
    methods: {
        createPopper () {
            if (typeof window === 'undefined') {
                return;
            }

            const modifiers = useBaseModifiers({ offset: this.offset });

            this.popperInstance = createPopper(
                (this as any).$refs.wrapper,
                (this as any).$refs.popup,
                {
                    strategy: 'fixed',
                    placement: this.placement as Placement,
                    modifiers,
                    ...this.popperOptions
                }
            );
        },
        destroyPopper () {
            if (this.popperInstance) {
                this.popperInstance.destroy();
                this.popperInstance = undefined;
            }
        }
    }
});
