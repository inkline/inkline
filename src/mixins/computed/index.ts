export * from '@inkline/inkline/mixins/computed/classes';
