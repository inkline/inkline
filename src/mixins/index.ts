export * from '@inkline/inkline/mixins/computed';
export * from '@inkline/inkline/mixins/props';
export { default as BroadcastDispatchMixin } from '@inkline/inkline/mixins/BroadcastDispatchMixin';
export { default as CollapsibleMixin } from '@inkline/inkline/mixins/CollapsibleMixin';
export { default as FormComponentMixin } from '@inkline/inkline/mixins/FormComponentMixin';
export { default as LinkableMixin } from '@inkline/inkline/mixins/LinkableMixin';
export { default as PopupMixin } from '@inkline/inkline/mixins/PopupMixin';
export { default as PopupControlsMixin } from '@inkline/inkline/mixins/PopupControlsMixin';
