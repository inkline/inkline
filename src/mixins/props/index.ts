export * from '@inkline/inkline/mixins/props/default';
export * from '@inkline/inkline/mixins/props/size';
