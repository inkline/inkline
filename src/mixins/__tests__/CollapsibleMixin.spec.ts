import { CollapsibleMixin } from '@inkline/inkline/mixins';
import { fireEvent, render } from '@testing-library/vue';
import { createMockInstance } from '@inkline/inkline/__mocks__/createMockInstance';

describe('mixins', () => {
    describe('CollapsibleMixin', () => {
        const Component = {
            name: 'Collapsible',
            mixins: [CollapsibleMixin],
            template: `<div :class="collapsibleClasses">
              <span>{{open}}</span>
              <button @click="setOpen(true)">Set open</button>
              <button @click="toggleOpen()">Toggle open</button>
            </div>`
        };

        it('should render correctly', () => {
            const wrapper = render(Component);

            expect(wrapper.html()).toMatchSnapshot();
        });

        describe('computed', () => {
            describe('classes', () => {
                it('should add classes based on props', () => {
                    const wrapper = render(Component, {
                        props: {
                            modelValue: true,
                            collapsible: true,
                            collapse: true
                        }
                    });

                    expect(wrapper.container.firstChild).toHaveClass(
                        '-open',
                        '-collapsible',
                        '-collapse-true'
                    );
                });
            });

            describe('collapsible', () => {
                it('should return true if collapse is true', () => {
                    const wrapper = createMockInstance(CollapsibleMixin, {
                        props: {
                            collapse: true
                        }
                    });

                    expect(wrapper.collapsible).toEqual(true);
                });

                it('should return false if collapse is false', () => {
                    const wrapper = createMockInstance(CollapsibleMixin, {
                        props: {
                            collapse: false
                        }
                    });

                    expect(wrapper.collapsible).toEqual(false);
                });

                it('should return whether window width is less than current breakpoint upper limit', () => {
                    const wrapper = createMockInstance(CollapsibleMixin, {
                        props: {
                            collapse: 'md'
                        }
                    });

                    wrapper.windowWidth = 800;
                    expect(wrapper.collapsible).toEqual(true);

                    wrapper.windowWidth = 1000;
                    expect(wrapper.collapsible).toEqual(false);
                });
            });
        });

        describe('methods', () => {
            describe('setOpen()', () => {
                it('should set open and emit update:modelValue', async () => {
                    const wrapper = render(Component);
                    const setOpenButton = await wrapper.getByText('Set open');

                    await fireEvent.click(setOpenButton);

                    expect(wrapper.emitted()['update:modelValue'][0]).toEqual([true]);
                    expect(wrapper.getByText('true')).toBeVisible();
                });
            });

            describe('toggleOpen()', () => {
                it('should toggle open and emit update:modelValue', async () => {
                    const wrapper = render(Component, {
                        props: { modelValue: true }
                    });
                    const toggleOpenButton = await wrapper.getByText('Toggle open');

                    await fireEvent.click(toggleOpenButton);

                    expect(wrapper.emitted()['update:modelValue'][0]).toEqual([false]);
                    expect(wrapper.getByText('false')).toBeVisible();
                });
            });

            describe('onWindowResize()', () => {
                it('should return if collapse is true', async () => {
                    const wrapper = createMockInstance(CollapsibleMixin, {
                        props: {
                            collapse: true
                        }
                    });

                    wrapper.setOpen = jest.fn();
                    wrapper.onWindowResize();

                    expect(wrapper.setOpen).not.toHaveBeenCalled();
                });

                it('should return if collapse is false', async () => {
                    const wrapper = createMockInstance(CollapsibleMixin, {
                        props: {
                            collapse: false
                        }
                    });

                    wrapper.setOpen = jest.fn();
                    wrapper.onWindowResize();

                    expect(wrapper.setOpen).not.toHaveBeenCalled();
                });

                it('should not setOpen to false if not transitioning to collapsed state', async () => {
                    const wrapper = createMockInstance(CollapsibleMixin, {
                        props: {
                            collapse: 'sm'
                        }
                    });

                    (global as any).innerWidth = 1024;
                    wrapper.setOpen = jest.fn();
                    wrapper.windowWidth = 800;
                    wrapper.onWindowResize();

                    expect(wrapper.setOpen).not.toHaveBeenCalled();
                });

                it('should setOpen to false if transitioning to collapsed state', async () => {
                    const wrapper = createMockInstance(CollapsibleMixin, {
                        props: {
                            collapse: 'md'
                        }
                    });

                    (global as any).innerWidth = 1024;
                    wrapper.setOpen = jest.fn();
                    wrapper.windowWidth = 800;
                    wrapper.onWindowResize();

                    expect(wrapper.setOpen).toHaveBeenCalledWith(false);
                });
            });
        });
    });
});
