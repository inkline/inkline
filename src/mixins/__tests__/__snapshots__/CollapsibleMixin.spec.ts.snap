// Vitest Snapshot v1

exports[`mixins > CollapsibleMixin > should render correctly 1`] = `"<div class=\\"-collapse-md\\"><span>false</span><button>Set open</button><button>Toggle open</button></div>"`;
